### ISSUE TITLE
Creating bootable USB devices from Windows 10 ISOs

### ISSUE THREAD

**Issue by mmstick** [pop-os/popsicle#67]
There's a request from time to time for the ability to flash USB devices with a Windows 10 ISO. Yet since Windows 10 ISOs are not hybrid, they require special setup.

1. Wipe the labels on the device
2. Format the device as NTFS
3. Mount the newly-formatted device
4. Mount the ISO
5. Extract the ISO's contents to the mounted USB device

**Comment by eduncan911** [comment]
I might be able to find time for this one.  Have to find a Windows machine first though...

I've scripted DISKPART multiple times to do this at many other companies.  I blogged the commands some time ago just so I could remember/script them:

https://eduncan911.com/blog/archives/create-a-bootable-usb-drive-with-windows.html

Only dangerous part is selecting the proper drive FIRST.  So, some script funsies is needed to detect removable drives and present options to the user as READLINE first.

How do you see a solution to this?  Would it be a BAT file the user would download and run?  Or, a copy-n-paste set of commands?  Maybe something the Installer would kick off?  (meaning, it needs to output some format for the GUI to be able to render options for, such as which flash drive to wipe/use)

**Comment by mmstick** [comment]
@eduncan911 There's no need for a Windows machine. You only need to verify that the flash drives are bootable once they've been written.

The steps listed above are what needs to be done. We can switch Popsicle to a Windows-flashing mode by mounting the ISO and checking if its contents are what's expected of a Windows ISO.

The Windows-flashing mode would differ from the regular mode by copying files in parallel to NTFS-partitioned, temporarily-mounted mount points, rather than writing bytes in parallel directly to the block devices.

We already have a crate for handling mounts of block devices and ISOs: [sys-mount](https://docs.rs/crate/sys-mount)

**Comment by eduncan911** [comment]
Why NTFS? Are there any files larger than 4GB that far32 couldn't support?

That way you don't have to worry about NTFS drivers.

**Comment by moriel5** [comment]
I forget, however I think that boot.wim may have gotten close to that size.

But I think that the issue is that Windows nowadays requires the main partition to be NTFS.

**Comment by olback** [comment]
To make this work, systems would need ntfs-3g installed? I might be wrong about this though and all that is needed is just a rust library that implements support for the NTFS filesystem.

**Comment by mmstick** [comment]
It could be added as an optional dependency, and the support enabled when this functionality is supported on the system.

**Comment by ids1024** [comment]
> I forget, however I think that boot.wim may have gotten close to that size.

According to the [documentation from Windows](https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/install-windows-from-a-usb-flash-drive#troubleshooting-file-copy-fails), it should be possible to boot the Windows installer from fat32. If `install.wim` is too large, it can be split into multiple files.

So that seems to be an option, if it proves easier.

**Comment by eduncan911** [comment]
I'm confused.  Are we talking about making a Windows 10 bootable USB?

I thought this request was to create a Pop_OS boot USB from *within* windows to *install* Pop_OS?

Why are we talking about `install.win`?  That's only for installing Windows.

Maybe I misunderstood those whole issue...  Lol

**Comment by ids1024** [comment]
> Maybe I misunderstood this whole issue... Lol

Seems so :wink:. As the title and description say, this is about "Windows 10 ISOs".

> I thought this request was to create a Pop_OS boot USB from within windows to install Pop_OS?

Support for running under Windows could be a good feature to add, but it is separate from what this is asking for.

**Comment by eduncan911** [comment]
Title says:

> Creating bootable USB devices ***from*** Windows 10 ISOs

It doesn't say, "Creating bootable Windows 10 USB devices."

Why would Pop_OS even care about creating a Windows 10 ISO?  Even more confused now.

(Btw, which I've done extensively with WinPE, and yes it does require a Windows machine or VM to pass USB through to).

**Comment by eduncan911** [comment]
Wait wait wait... Are we talking about some script or app that runs from DE, e.g. Gnome, that can create Windows USB boot devices?

Ah...  I think I finally get it.  I guess I didn't undwrstabd what "Popsicle" was.

**Comment by ids1024** [comment]
> Ah... I think I finally get it. I guess I didn't understand what "Popsicle" was. (I thought it was to install Pop_os)

It's the program that appears under the name "USB Flasher", which comes with Pop!_OS.

> If this is a scripting/binary/python/go/ruby thing, then yeah I can still do it.
> 
> If this is an App thing, someone else needs to as I have zero app dev experience on Linux/gnome - outside of console UI that is. (E.g. VLC's CLI UX)

Take a look at the repo this issue is posted on. It's written in Rust, split into a library, with both a CLI and GTK frontend. So maybe not your choice of language, but the key part of implementing this is probably adding code to the library; not the GUI.

**Comment by TeamLinux01** [comment]
A little off topic, but I figured it was related. I have been using a program call Ventoy for all my Linux/Windows booting needs now.

https://ventoy.net/en/index.html

Don't know if something like this would be the route to go in the future.

**Comment by ibnYusrat** [comment]
So.. does popsicle support Flashing Windows ISOs by now?

**Comment by Heroofthenexus** [comment]
I hope this feature will be added in soon, it's sorely needed.

**Comment by techguy16** [comment]
Hi, have a look at the 'windows-iso' branch. This branch adds support for flashing Windows ISOs from Popsicle. The branch has existed for about a   year so thr version of Popsicle is a little bit old but it still works.

**Comment by dginovker** [comment]
Fwiw, Impression (https://flathub.org/apps/io.gitlab.adhami3310.Impression) seems to support Windows while we "patiently" wait here ^^

**Comment by moriel5** [comment]
@techguy16 Indeed. I'm using WoeUSB and a spare Windows 10/11 installation with Rufus in the meantime, however it is getting tiring installing GRUB2 just for WoeUSB and then removing both after flashing the ISO, or going to another machine each time.
